cmd/openfga/run.go

```diff
@@ -0,0 +1,773 @@
+			Enabled: true,
+			Port:    3000,
+		},
+		Profiler: ProfilerConfig{
+			Enabled: false,
+			Addr:    ":3001",
+		},
+	}
+}
+
+func DefaultConfigWithRandomPorts() *Config {
+	config := DefaultConfig()
+
+	// Since this is used for tests, turn the following off:
+	config.Playground.Enabled = false
+
+	l, err := net.Listen("tcp", "")
+	if err != nil {
+		panic(err)
+	}
+	defer l.Close()
+	httpPort := l.Addr().(*net.TCPAddr).Port
+
+	l, err = net.Listen("tcp", "")
+	if err != nil {
+		panic(err)
+	}
+	defer l.Close()
+	grpcPort := l.Addr().(*net.TCPAddr).Port
+
+	config.GRPC.Addr = fmt.Sprintf("0.0.0.0:%d", grpcPort)
+	config.HTTP.Addr = fmt.Sprintf("0.0.0.0:%d", httpPort)
+
+	return config
+}
+
+// ReadConfig returns the OpenFGA server configuration based on the values provided in the server's 'config.yaml' file.
+// The 'config.yaml' file is loaded from '/etc/openfga', '$HOME/.openfga', or the current working directory. If no configuration
+// file is present, the default values are returned.
+func ReadConfig() (*Config, error) {
+	config := DefaultConfig()
+
+	viper.SetConfigName("config")
+	viper.SetConfigType("yaml")
+
+	configPaths := []string{"/etc/openfga", "$HOME/.openfga", "."}
+	for _, path := range configPaths {
+		viper.AddConfigPath(path)
+	}
+	viper.SetEnvPrefix("OPENFGA")
+	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
+	viper.AutomaticEnv()
+
+	err := viper.ReadInConfig()
+	if err != nil {
+		if _, ok := err.(viper.ConfigFileNotFoundError); !ok {
+			return nil, fmt.Errorf("failed to load server config: %w", err)
+		}
+	}
+
+	if err := viper.Unmarshal(config); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal server config: %w", err)
+	}
+
+	return config, nil
+}
+
+func VerifyConfig(cfg *Config) error {
+	if cfg.ListObjectsDeadline > cfg.HTTP.UpstreamTimeout {
+		return fmt.Errorf("config 'http.upstreamTimeout' (%s) cannot be lower than 'listObjectsDeadline' config (%s)", cfg.HTTP.UpstreamTimeout, cfg.ListObjectsDeadline)
+	}
+
+	if cfg.Playground.Enabled {
+		if !cfg.HTTP.Enabled {
+			return errors.New("the HTTP server must be enabled to run the openfga playground")
+		}
+
+		if !(cfg.Authn.Method == "none" || cfg.Authn.Method == "preshared") {
+			return errors.New("the playground only supports authn methods 'none' and 'preshared'")
+		}
+	}
+
+	if cfg.HTTP.TLS.Enabled {
+		if cfg.HTTP.TLS.CertPath == "" || cfg.HTTP.TLS.KeyPath == "" {
+			return errors.New("'http.tls.cert' and 'http.tls.key' configs must be set")
+		}
+	}
+
+	if cfg.GRPC.TLS.Enabled {
+		if cfg.GRPC.TLS.CertPath == "" || cfg.GRPC.TLS.KeyPath == "" {
+			return errors.New("'grpc.tls.cert' and 'grpc.tls.key' configs must be set")
+		}
+	}
+
+	return nil
+}
+
+func run(_ *cobra.Command, _ []string) {
+	config, err := ReadConfig()
+	if err != nil {
+		panic(err)
+	}
+
+	if err := runServer(context.Background(), config); err != nil {
+		panic(err)
+	}
+}
+
+func runServer(ctx context.Context, config *Config) error {
```

Comment:
Let's do a sanity check on the usage of `ctx` in this method body. I see a couple of spots where we don't propagate the provided context.

Reply:
There are two spots that don't use the passed in context: one is where we use it as a timeout for a `grpc.Dialcontext`, and the other is for a shutdown context.